pkg_resources/tests/test_pkg_resources.py

```diff
@@ -425,3 +426,56 @@ def test_normalize_path_backslash_sep(self, unnormalized, expected):
         """Ensure path seps are cleaned on backslash path sep systems."""
         result = pkg_resources.normalize_path(unnormalized)
         assert result.endswith(expected)
+
+
+class TestWorkdirRequire:
+    def fake_site_packages(self, tmp_path, monkeypatch, dist_files):
+        site_packages = tmp_path / "site-packages"
+        site_packages.mkdir()
+        for file, content in self.FILES.items():
+            path = site_packages / file
+            path.parent.mkdir(exist_ok=True, parents=True)
+            path.write_text(inspect.cleandoc(content), encoding="utf-8")
+
+        monkeypatch.setattr(sys, "path", [site_packages])
+        return os.fspath(site_packages)
+
+    FILES = {
+        "pkg1_mod-1.2.3.dist-info/METADATA": """
+            Metadata-Version: 2.4
+            Name: pkg1.mod
+            Version: 1.2.3
+            """,
+        "pkg2.mod-0.42.dist-info/METADATA": """
+            Metadata-Version: 2.1
+            Name: pkg2.mod
+            Version: 0.42
+            """,
+        "pkg3_mod.egg-info/PKG-INFO": """
+            Name: pkg3.mod
+            Version: 1.2.3
+            """,
+        "pkg4.mod.egg-info/PKG-INFO": """
+            Name: pkg4.mod
+            Version: 0.42
+            """,
+    }
+
+    @pytest.mark.parametrize(
+        ("name", "version", "req"),
+        [
+            ("pkg1.mod", "1.2.3", "pkg1.mod>=1"),
+            ("pkg2.mod", "0.42", "pkg2.mod>=0.4"),
+            ("pkg3.mod", "1.2.3", "pkg3.mod<=2"),
+            ("pkg4.mod", "0.42", "pkg4.mod>0.2,<1"),
+        ],
+    )
+    def test_require_normalised_name(self, tmp_path, monkeypatch, name, version, req):
+        # https://github.com/pypa/setuptools/issues/4853
+        site_packages = self.fake_site_packages(tmp_path, monkeypatch, self.FILES)
+        ws = pkg_resources.WorkingSet([site_packages])
+
+        [dist] = ws.require(req)
+        assert dist.version == version
+        assert dist.project_name == name
```

Comment:
```python
assert dist.project_name == name
```

Sorry apparently this expectation was misleading from my part.

Reply:
Yeah, looks like underscores are already being converted to dashes for `project_name` here vis `safe_name` prior to this change. Do you want to assert on that instead?

Reply:
I think we can simply remove this assert. We possibly can have slightly different versions for each mock distribution so that they don't overlap, but not fundamental.

Reply:
Dropped it in 65e078f like you did in 2c242238f536c4b942812632ba9dd0b1c48b4b85.